h3-quinn/src/lib.rs

```diff
@@ -1,223 +1,149 @@
+/// A [`quic::Connection`] backed by [`quinn::Connection`].
 pub struct Connection {
     conn: quinn::Connection,
-    incoming_bi: IncomingBiStreams,
-    opening_bi: Option<OpenBi>,
-    incoming_uni: IncomingUniStreams,
-    opening_uni: Option<OpenUni>,
 }
 
 impl Connection {
-    /// Create a [`Connection`] from a [`quinn::NewConnection`]
-    pub fn new(new_conn: NewConnection) -> Self {
-        let NewConnection {
-            uni_streams,
-            bi_streams,
-            connection,
-            ..
-        } = new_conn;
-
-        Self {
-            conn: connection,
-            incoming_bi: bi_streams,
-            opening_bi: None,
-            incoming_uni: uni_streams,
-            opening_uni: None,
-        }
+    /// Create a [`Connection`] from a [`quinn::Connection`]
+    pub fn new(conn: quinn::Connection) -> Self {
+        Self { conn }
     }
 }
 
-/// The error type for [`Connection`]
-///
-/// Wraps reasons a Quinn connection might be lost.
-#[derive(Debug)]
-pub struct ConnectionError(quinn::ConnectionError);
-
-impl std::error::Error for ConnectionError {}
-
-impl fmt::Display for ConnectionError {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.0.fmt(f)
-    }
-}
-
-impl Error for ConnectionError {
-    fn is_timeout(&self) -> bool {
-        matches!(self.0, quinn::ConnectionError::TimedOut)
-    }
-
-    fn err_code(&self) -> Option<u64> {
-        match self.0 {
-            quinn::ConnectionError::ApplicationClosed(quinn_proto::ApplicationClose {
-                error_code,
-                ..
-            }) => Some(error_code.into_inner()),
-            _ => None,
-        }
-    }
-}
-
-impl From<quinn::ConnectionError> for ConnectionError {
-    fn from(e: quinn::ConnectionError) -> Self {
-        Self(e)
-    }
-}
-
-impl<B> quic::Connection<B> for Connection
-where
-    B: Buf,
-{
+impl<B: Buf> quic::Connection<B> for Connection {
+    type OpenStreams = OpenStreams;
+    type BidiStream = BidiStream<B>;
     type SendStream = SendStream<B>;
     type RecvStream = RecvStream;
-    type BidiStream = BidiStream<B>;
-    type OpenStreams = OpenStreams;
     type Error = ConnectionError;
 
     fn poll_accept_bidi(
         &mut self,
-        cx: &mut task::Context<'_>,
+        cx: &mut Context<'_>,
     ) -> Poll<Result<Option<Self::BidiStream>, Self::Error>> {
-        let (send, recv) = match ready!(self.incoming_bi.next().poll_unpin(cx)) {
-            Some(x) => x?,
-            None => return Poll::Ready(Ok(None)),
-        };
-        Poll::Ready(Ok(Some(Self::BidiStream {
-            send: Self::SendStream::new(send),
-            recv: Self::RecvStream::new(recv),
-        })))
+        Poll::Ready(
+            match ready!(Box::pin(self.conn.accept_bi()).poll_unpin(cx)) {
+                Ok((send, recv)) => Ok(Some(Self::BidiStream::new(
+                    Self::SendStream::new(send),
+                    Self::RecvStream::new(recv),
+                ))),
```

Comment:
Note that there is a situation that returns `Ready(Ok(None))`, and checked inside `IncomingStream::poll_next` before, which is when the error is `ConnectionError::LocallyClosed`. 

We may choose to handle that in `h3` directly though.

Reply:
Adding that fixes everything else except `tests::connection::control_close_send_error`.